### ISSUE TITLE
CI: Add proper testing for distributed dataflows

### ISSUE THREAD

**Issue by phil-opp** [dora-rs/dora#475]
Our current CI test runs the coordinator and two daemons on the same machine. To ensure that dora is able to communicate across machines, we would need to use three different machines with different IPs. I'm not sure if something like this is possible on GitHub actions, so this issue is mostly a reminder to look into this further.

Notes:

- @haixuanTao mentioned the https://github.com/haythem/public-ip in https://github.com/dora-rs/dora/pull/471#discussion_r1568389104

**Comment by Michael-J-Ward** [comment]
As far as I can tell, the `examples` runners in CI serve as `dora`'s integration tests. Run Completes = Green, but nothing verifies the output of the run - is that correct?

Are any of them deterministic enough that we could throw a `snapshot` tests on the output / logs - or otherwise test "successful end state"?

**Comment by phil-opp** [comment]
Yes, we don't very the output, as it's not deterministic enough. We just check the exit code of the process, which should be `!= 0` when an error occurs. We also have some examples with assertions, for example for verifying that the ROS2 turtle is at a different position at the end of the test (to ensure that it moved).

**Comment by heyong4725** [comment]
Closing — folded into #1800 (1.0 CI coverage backlog), specifically bucket B1: "Add an E2E that brings up a multi-daemon cluster, starts a dataflow, records, stops, replays from the recording, validates the replayed state."

That bucket is now the single live tracker for this work and includes the broader context (Cluster lifecycle smoke and Record/replay round-trip exist as separate nightly jobs today; the missing piece is one E2E that stitches them together with state validation).

The 3-different-IPs angle from the OP is harder on GitHub-hosted runners but worth keeping in mind when the test is implemented.
